### PR TITLE
Redshift: add [ALTER/CREATE/DROP] PROCEDURE segments

### PR DESCRIPTION
**Pull request by rpr-ableton**
### Brief summary of the change made

What the title says :slightly_smiling_face: 

Adding relevant segments for procedures. It diverges slightly from postgresql due to the additional optional `argmode` in the parameters list.

Some more changes done along the way (cleanup, renaming etc):
- move CompressionType, ArgMode and ColumnEncoding to grammars instead of segments
- clean up StatementSegment to keep only statements
- renaming some classes for the sake of clarity

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by barrywhart** [comment]
@rpr-ableton: One of the tests is failing with a parse error. Can you look into it?

**Comment by rpr-ableton** [comment]
Hey @barrywhart, for sure! I naively assumed that everything would be good to go since the yaml generated from `generate_parse_fixtures` looked good :sweat_smile:

I have some time set aside for sqlfluff this afternoon.

**Review by tunetheweb — CHANGES_REQUESTED** [review]

**Comment by rpr-ableton** [comment]
@tunetheweb i've adopted the changes you suggested, thanks for that! Let me know if you see space for any further cleanup while we're at it :smile: 

> This seems to be causing the failure. Weirdly it works running the generate in python with this syntax, but not within tox (whether within GitHub Action or even on my own machine). Don't understand why as it should be the same. Very odd...

I'm not sure what to do about that one tbh. Should I try to find another way to express this logic?

**Comment by tunetheweb** [comment]
This seems to work:

```py
    type = "procedure_parameter_list"
    # Odd syntax, but prevents eager parameters being confused for data types
    _param_type = OneOf("REFCURSOR", Ref("DatatypeSegment"))
    match_grammar = Bracketed(
        Sequence(
            OneOf(Ref("ParameterNameSegment"),exclude=_param_type,optional=True),
            Ref("ArgModeSegment",optional=True),
            _param_type,
            AnyNumberOf(
                Sequence(
                    Ref("CommaSegment"),
                    OneOf(Ref("ParameterNameSegment"),exclude=_param_type,optional=True),
                    Ref("ArgModeSegment",optional=True),
                    _param_type,
                ),
            ),
            optional=True,
        ),
    )
```

File: src/sqlfluff/dialects/dialect_redshift.py [hunk 1]
```diff
@@ -1193,6 +1175,106 @@ class CreateSchemaStatementSegment(BaseSegment):
...
+                        Ref("ParameterNameSegment"),
+                        min_times=1,
+                    ),
+                    OneOf("REFCURSOR", Ref("DatatypeSegment")),
+                ),
+                OneOf("REFCURSOR", Ref("DatatypeSegment")),
+            ),
+            optional=True,
+        ),
+    )
```
Comment (tunetheweb) [inline src/sqlfluff/dialects/dialect_redshift.py:1202]:
```suggestion
    type = "procedure_parameter_list"
    # Odd syntax, but prevents eager parameters being confused for data types
    _param_type = OneOf("REFCURSOR", Ref("DatatypeSegment"))
    match_grammar = Bracketed(
        Sequence(
            OneOf(Ref("ParameterNameSegment"),exclude=_param_type,optional=True),
            Ref("ArgModeSegment",optional=True),
            _param_type,
            AnyNumberOf(
                Sequence(
                    Ref("CommaSegment"),
                    OneOf(Ref("ParameterNameSegment"),exclude=_param_type,optional=True),
                    Ref("ArgModeSegment",optional=True),
                    _param_type,
                ),
            ),
            optional=True,
        ),
    )
```

**Comment by rpr-ableton** [comment]
That doesn't quite work as `ParameterNameSegment` and `ArgModeGrammar` can be swapped around in position.

Something like this should do the trick:

```    type = "procedure_parameter_list"
    # Odd syntax, but prevents eager parameters being confused for data types
    _param_type = OneOf("REFCURSOR", Ref("DatatypeSegment"))
    match_grammar = Bracketed(
        Sequence(
            AnyNumberOf(
                OneOf(Ref("ParameterNameSegment"),exclude=_param_type),
                Ref("ArgModeGrammar"),
                optional=True,
            ),
            _param_type,
            AnyNumberOf(
                Sequence(
                    Ref("CommaSegment"),
                    AnyNumberOf(
                        OneOf(Ref("ParameterNameSegment"),exclude=_param_type),
                        Ref("ArgModeGrammar"),
                        optional=True,
                    ),
                    _param_type,
                ),
            ),
            optional=True,
        ),
    )```

**Comment by tunetheweb** [comment]
What about this slight variation?

```py
    type = "procedure_parameter_list"
    # Odd syntax, but prevents eager parameters being confused for data types
    _param_type = OneOf("REFCURSOR", Ref("DatatypeSegment"))
    match_grammar = Bracketed(
        Sequence(
            AnyNumberOf(
                Ref("ParameterNameSegment"),
                Ref("ArgModeSegment"),
                exclude=_param_type,
                max_times_per_element=1,
            ),
            _param_type,
            AnyNumberOf(
                Sequence(
                    Ref("CommaSegment"),
                    AnyNumberOf(
                        Ref("ParameterNameSegment"),
                        Ref("ArgModeSegment"),
                        exclude=_param_type,
                        max_times_per_element=1,
                    ),
                    _param_type,
                ),
            ),
            optional=True,
        ),
    )
```

Testing it now and then will commit so it makes 0.11.0 which I'm about to release.

**Comment by tunetheweb** [comment]
Woo hoo that worked. Presume you're happy to merge this @rpr-ableton ?

**Review by tunetheweb — APPROVED** [review]

**Comment by rpr-ableton** [comment]
LGTM, thanks!

**Pull request merged**